### PR TITLE
Added option for firing and rotating the ship with mouse

### PR DESCRIPTION
**Pull request by CAPTAIN1947**
**Feature:**

## Feature Details
This adds a new option for flying the ship with mouse. The right mouse buttons can be used for firing the primary weapons. And rotating the mouse around the ship rotates the ship towards the mouse maintaining the steering capabilities of the ship. Hopefully it will improve the experience for those who prefer to fight rather than avoiding it. 
Also, its another step towards #955.

## UI Screenshots
Before:
image
After:
image

## Usage Examples
It allows more precise firing. As with the previous keyboard rotation it was possible to miss a small ship, especially at low zoom as often pressing the LEFT key turned the ship too left and vice-versa for the RIGHT key. Also, it creates future possibility to add  controller support for the game, allowing it to be ported to platforms like Steam Deck. Also, it may be useful to people with special needs and can be used as an accessibility feature. 

## Testing Done
I completed some missions with this feature and didn't face any serious bugs. 

## Performance Impact
I shouldn't affect the performance to any great extent, though I haven't made any benchmarks.

### CODE REVIEW

**Comment by CAPTAIN1947** [comment]
Currently the ship with mouse controls enabled the ship is unable to board a disabled ship properly.

**Comment by CAPTAIN1947** [comment]
After some discussion in the discord now the mouse-turning can be toggled like fast-forward with left-alt as the default binding for it.

**Comment by Hurleveur** [comment]
I like the alt triggering this.
That said I think turrets should point to the point you're aiming at as well, as fast as they can; as if it was a target.

**Comment by CAPTAIN1947** [comment]
Thanks for the feedback...
> That said I think turrets should point to the point you're aiming at as well, as fast as they can; as if it was a target. >
I will implement it asap.

**Comment by CAPTAIN1947** [comment]
After some recent discussions in discord maybe the changes related to turrets that @Hurleveur suggested can be done with #5459 as that will change the way turrets work and bring new possibilities.

**Comment by Hurleveur** [comment]
That other PR is big enough already as it is in my opinion, but it could be a follow up PR after this one and that turret angle PR is merged too.

File: source/AI.cpp
Comment (Hurleveur) [inline source/AI.cpp:500]: I do not know much about commands but can't the moveAngle and the rightMouseButtonHeld be added to the command?
rightMouseButtonHeld is just to know if we are shooting, right? The idea of the command is that the game engine parses meaning out of whatever the player is typing (not only but it's one thing)

File: source/AI.cpp
Comment (Hurleveur) [inline source/AI.cpp:3559]: I feel like this logic can be compacted but I'm not really sure how exactly ; just giving a direction that matches the difference between the two angles and not checking every single case
I'm not sure how much the fact that these degrees are between 0 and 360 instead of -180 and 180 is helping for that either.

File: source/Angle.h
Comment (Hurleveur) [inline source/Angle.h:53]: why do we need to convert it this way?

**Review by Hurleveur — COMMENTED** [review]
I really like the way this works, I'm just not sure if this is the best way to do it
RightbuttonHeld should not be passed to Step and instead Engine should determine if we need to shoot or not with that button held; meaning it needs to know if we are in that state (idk how it works with the other modes like fast forward but we should do the same probably) and then you can just have a getter if one needs to know the state and doesn't have it.

File: source/AI.cpp
Comment (CAPTAIN1947) [reply source/AI.cpp:500]: Well, I also felt that what I did was not the best way to do it... But couldn't work out a better way. @quyykk advised to avoid calling SDL methods in the AI.cpp itself. And what I saw from existing commands, none of them handles the mouse related things. Also, the other mouse related operations like drag, click, etc are hard coded in the events file. I too wanted to make the system more dynamic so that it will easy in the future to map mouse keys, etc. If you can show me a better way I will be glad to do it in your way.

File: source/Angle.h
Comment (CAPTAIN1947) [reply source/Angle.h:53]: I am not really sure about this one... Just copied the way the Degrees() function was done.

**Comment by CAPTAIN1947** [comment]
I have my exams from next week. So, there will be some delay. I will again be working on it after that.

**Comment by CAPTAIN1947** [comment]
I am back and I made the logic related to the angle checking much more shorter(4 lines from 17) and also removed the Degrees360 function that I created... as yeah @Hurleveur was right... it doesn't makes much difference if the angle is between -180 to 180 or 0 to 360. I could have made the angle checking line more smaller; I did it in one line first... but that makes the line too long and complex. Also, as I was using the difference of the angles a lot of times so fell that it would be better to cache it.

File: source/AI.cpp
Comment (CAPTAIN1947) [reply source/AI.cpp:500]: Now onto this one... Yeah thanks for that advice. I think that the commands infrastructure will help me on this.

Comment (CAPTAIN1947) [reply source/AI.cpp:500]: I have made it, so that it uses the PRIMARY command but there is a big problem. The primary command is a general command for the ship to fire its primary weapons... there is no way I can tell if that command is generated after a mouse input or keyboard. So, basically it ain't possible to make the weapons fire with right mouse button only if mouse turning is enabled without using any extra flags.

**Review by samrocketman — CHANGES_REQUESTED** [review]
# Feedback

This feedback is organized into three major sections.

* Code style fixes
* Good experiences
* Should probably change


# Code style fixes

Apply the following patch for code style fixes.

    git apply style-fix.patch

<details><summary><b>Here is the patch file</b> (Click to expand)</summary>

---


Here is `style-fix.patch`.

```diff
diff --git a/source/AI.cpp b/source/AI.cpp
index d4b4c39a..8fa306b6 100644
--- a/source/AI.cpp
+++ b/source/AI.cpp
@@ -3518,14 +3518,15 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
 			&& !isMouseTurningEnabled)
 		AutoFire(ship, firingCommands, false);
 
-	if (activeCommands.Has(Command::MOUSETURNING)) {
+	if(activeCommands.Has(Command::MOUSETURNING))
+	{
 		isMouseTurningEnabled = !isMouseTurningEnabled;
 	}
 
-	if (isMouseTurningEnabled && !ship.IsBoarding() && !ship.IsReversing())
+	if(isMouseTurningEnabled && !ship.IsBoarding() && !ship.IsReversing())
 	{
 		Angle shipAngle = ship.Facing();
-		if (rightMouseButtonHeld)
+		if(rightMouseButtonHeld)
 		{
 			int index = 0;
 			for(const Hardpoint &hardpoint : ship.Weapons())
@@ -3537,7 +3538,7 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
 		}
 
 		double angDiff = mouseAngle.Degrees() - shipAngle.Degrees();
-		if (abs(angDiff) > 1)
+		if(abs(angDiff) > 1)
 			command.SetTurn((angDiff / abs(angDiff)) * pow(-1, (floor(abs(angDiff) / 180))));
 		else
 			command.SetTurn(0);
@@ -3586,8 +3587,7 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
 	}
 	bool shouldAutoAim = false;
 	if(Preferences::Has("Automatic aiming") && !command.Turn() && !ship.IsBoarding()
-			&& ((target && target->GetSystem() == ship.GetSystem() && target->IsTargetable())
-				|| ship.GetTargetAsteroid())
+			&& ((target && target->GetSystem() == ship.GetSystem() && target->IsTargetable()) || ship.GetTargetAsteroid())
 			&& !autoPilot.Has(Command::LAND | Command::JUMP | Command::FLEET_JUMP | Command::BOARD)
 			&& !isMouseTurningEnabled)
 	{
diff --git a/source/Engine.cpp b/source/Engine.cpp
index 24da66fd..c29c0ddf 100644
--- a/source/Engine.cpp
+++ b/source/Engine.cpp
@@ -22,9 +22,6 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 #include "FillShader.h"
 #include "Fleet.h"
 #include "Flotsam.h"
-#include "text/Font.h"
-#include "text/FontSet.h"
-#include "text/Format.h"
 #include "FrameTimer.h"
 #include "GameData.h"
 #include "Government.h"
@@ -38,7 +35,6 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 #include "NPC.h"
 #include "OutlineShader.h"
 #include "Person.h"
-#include "pi.h"
 #include "Planet.h"
 #include "PlanetLabel.h"
 #include "PlayerInfo.h"
@@ -61,6 +57,10 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 #include "TestContext.h"
 #include "Visual.h"
 #include "Weather.h"
+#include "pi.h"
+#include "text/Font.h"
+#include "text/FontSet.h"
+#include "text/Format.h"
 #include "text/WrappedText.h"
 
 #include <algorithm>
@@ -2139,23 +2139,23 @@ void Engine::DoWeather(Weather &weather)
 
 void Engine::HandleMouseInput(Command &activeCommands)
 {
-	if (activeCommands.Has(Command::MOUSETURNING))
+	if(activeCommands.Has(Command::MOUSETURNING))
 		isMouseTurningEnabled = !isMouseTurningEnabled;
-	if (isMouseTurningEnabled)
+	if(isMouseTurningEnabled)
 	{
 		int mousePosX;
 		int mousePosY;
-		if ((SDL_GetMouseState(&mousePosX, &mousePosY) & SDL_BUTTON_RMASK) != 0)
+		if((SDL_GetMouseState(&mousePosX, &mousePosY) & SDL_BUTTON_RMASK) != 0)
 			rightMouseButtonHeld = true;
 		else
 			rightMouseButtonHeld = false;
-		double relx = mousePosX - Screen::RawWidth()/2;
-		double rely = mousePosY - Screen::RawHeight()/2;
-		if (relx == 0)
+		double relx = mousePosX - Screen::RawWidth() / 2;
+		double rely = mousePosY - Screen::RawHeight() / 2;
+		if(relx == 0)
 			mouseAngle = 90;
 		else
-			mouseAngle = (180/PI)*(atan(rely/relx)) + 90;
-		if (relx < 0)
+			mouseAngle = (180 / PI) * (atan(rely / relx)) + 90;
+		if(relx < 0)
 			mouseAngle += 180;
 	}
 }
```

</details>

---

# Good experiences

Controlling the fighter with a mouse (toggling with left alt) works as advertised.

Mining without an asteroid scanner is really easy.  Targeting ships and changing ship targets without focusing on the ship as a target is also really easy.  This drastically improves how nimble a player can be with a ship.

Playing with double speed is also easier because the ship moves at the rate wherever the player points their mouse.  So double time improves the accuracy in which a player points their ship or guns.  I think this is also an improvement.

# Should probably change

### Reloading game

When loading into a new game or reloading the game after death.  The mouse control is disabled.  Even if the player toggled it on the mouse control is still disabled.

You have to toggle mouse control again to turn it on; that's not a big deal.  It should probably change for players who'd want to use it full time.

### Preferences UI (optional feedback)

The preference UI looks okay but it does wrap.  I'll think about how we can
improve this.  Wrapping makes preferences a little less clear.

<details><summary><b>Screenshot of preference wrapping</b> (Click to expand)</summary>

---

Screenshot taken at 2022-07-04_10-51-25

</details>

---

A separate pull request after this one gets merged would be good to fix the preferences UI.

On the left side we could have: fleet, weapons, targeting.  And on the right side of the pane: Navigation and Interface.  Reorganizing this later would prevent preference wrapping but it will likely cause a bunch of conflicts for other pull requests which is why I want it to be a separate PR after this one gets merged.

### Fleet combat

* Do not move fleet when right clicking (to fire)
* Issue fleet target when right clicking on a targetable object or ship so that your fleet does attack it.

When playing with more than one ship and mouse control is togggled on, you can't have both a fleet and use your ship's guns.  By right clicking for the flagship to fire your fleet will travel and hold station at a location.

Right clicking to move fleet should probably be disabled for this unless a target is right clicked.  For example, a player right clicks on a ship your fleet should probably still focus fire on a ship you're attacking.

The only place you'd need to change to address this feedback would be in Engine.cpp

https://github.com/endless-sky/endless-sky/blob/e17f078f106cf61f14ea0b98766791f2375bc54f/source/Engine.cpp#L1958-L1959

File: source/AI.cpp
Comment (Hurleveur) [reply source/AI.cpp:500]: You don't need to know how it was generated; AI.cpp should just executes commands when its possible, in my opinion.
That means Engine.cpp knows if its mouse or keyboard and if its mouse and we're in alt mode then it gives a primary command, otherwise not.

**Comment by samrocketman** [comment]
I have opened a pull request on @CAPTAIN1947 for https://github.com/CAPTAIN1947/endless-sky/pull/1.  I implemented all of my suggested feedback for them and they need only to merge it.

File: source/AI.cpp
Comment (samrocketman) [inline source/AI.cpp:3536]: Probably better to set the firing command instead of iterate the hardpoints.  This is a copy of existing code.

File: source/AI.cpp
Comment (samrocketman) [reply source/AI.cpp:500]: @Hurleveur would it make sense to support `Angle` within `PlayerInfo &player`?  The right click held can set a firing command instead of being passed as a step.  If both of these are done then the Step arguments can be reverted.

For example,

```cpp
player.GetMouseAngle()
```

File: source/AI.cpp
Comment (samrocketman) [inline source/AI.cpp:3522]: This is no longer required if you rely on the preference `Preferences::Has("alt-mouse turning")` which I introduced in my feedback pull request to your fork.  https://github.com/CAPTAIN1947/endless-sky/pull/1

**Comment by Hurleveur** [comment]
> @Hurleveur would it make sense to support Angle within PlayerInfo &player? The right click held can set a firing command instead of being passed as a step. If both of these are done then the Step arguments can be reverted.

I think we can justify it given it is the angle of the player's flagship.
Good idea, that way we give the last angle to it and pass it over to AI without AI having to call anything from the Engine which could look weird.

**Comment by samrocketman** [comment]
https://github.com/CAPTAIN1947/endless-sky/pull/1 is all of the feedback addressed.

**Comment by CAPTAIN1947** [comment]
marking all the feedback as addressed. :)

File: source/Engine.cpp
Comment (Hurleveur) [inline source/Engine.cpp:2161]: I think it would be good to add a comment somewhere around here to explain whats going on

**Review by Hurleveur — DISMISSED** [review]
Looks good to me now!
AI's function is no longer modified, the way to do it by pressing alt is nice, all I can think of is maybe someone could want to add more features into it, but it seems to go to for me (and can be done in another PR if needed).
9.1.5 being steam-deck compatible would be nice! (that said, we would need reviews from main devs for that)

File: source/Engine.cpp
Comment (CAPTAIN1947) [reply source/Engine.cpp:2161]: ok will do that

Comment (CAPTAIN1947) [reply source/Engine.cpp:2165]: Done!

**Review by samrocketman — APPROVED** [review]
Following up on previous review.  This plays cleanly and is a necessity for playing ES on the steam deck.

Early game is nice with single ship control as well.

**Review by Hurleveur — APPROVED** [review]
Review was dismissed but all that changed was more comments, so I'm still 100% for this to be merged

**Comment by Hurleveur** [comment]
Could you please update this with master? I like to play on this branch x)

**Comment by samrocketman** [comment]
> done... not sure if the logic will still work or not. Good to see that you are still using it XD

It does still work.  I have included this in my fork and play it daily.  If there was an issue then I would have opened a bugfix PR after debugging.

As far as I can tell this change has been bug free game with around 100+ hours of playtesting both on PC and steam deck.

**Comment by Hurleveur** [comment]
Same, tested about 50hours without issue

**Comment by CAPTAIN1947** [comment]
Glad to hear that! Enjoy.

**Comment by CAPTAIN1947** [comment]
Done

File: source/AI.cpp
Comment (warp-core) [inline source/AI.cpp:3716]: Since it looks like the mouse angle isn't actually important until this point, how about just storing the mouse position as a Point (relative to the centre of the screen) and doing a dot product here to figure out how well aligned the ship is to the mouse vector?



```suggestion

	if(Preferences::Has("alt-mouse turning") && !ship.IsBoarding() && !ship.IsReversing())
	{
		Point shipAngle = ship.Unit();

		double angDiff = player.MouseAngle().Unit().Dot(shipAngle);
		double mag = abs(angDiff);
		if(mag > 1.)
			command.SetTurn(angDiff > 0. ? 1. : -1.);
		else
			command.SetTurn(0);
	}

```

`player.MouseAngle().Unit()` should probably become `player.MouseUnit()`, and just store the unit vector in the direction of hte cursor.

**Review by warp-core — COMMENTED** [review]
The Command name should probably have an `_` separating hte two words.

**Comment by warp-core** [comment]
So the targeting options are still lined up with the modifier description box.

image

**Review by warp-core — APPROVED** [review]

**Comment by CAPTAIN1947** [comment]
I sorry I couldn't respond earlier as I was a bit busy lately and thanks for the review. Please tell me what else needs to be done for getting this merged?

**Comment by CAPTAIN1947** [comment]
> So the targeting options are still lined up with the modifier description box.
> 
> image

Well this is one of the possible solutions I too thought about but I am not sure if it is a good idea to have two navigation tabs.

**Comment by warp-core** [comment]
> I sorry I couldn't respond earlier as I was a bit busy lately and thanks for the review. Please tell me what else needs to be done for getting this merged?

You don't need to do anything more.

> Well this is one of the possible solutions I too thought about but I am not sure if it is a good idea to have two navigation tabs. But if the other devs are okay with it then I don't have any problem.

I think this is better than putting the new entry where the other navigation control options are.
This panel could clearly do with some improvements, but I think this is a good enough solution for now. I don't think it looks particularly bad and it's not like it'll make future improvements more difficult.

**Pull request merged**
